File: client/src-tauri/src/commands/session.rs
// MIT License
// Copyright (c) 2026 190615273@qq.com
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

//! 会话管理命令（HTTP-fallback）
//!
//! 前端通过 HTTP `/v1/sessions/*` 优先调用，失败后降级至此 IPC 通道。

use crate::Message;
use crate::Session;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;
use tauri::State;
use tracing::info;
use uuid::Uuid;

pub struct AppState {
    pub sessions: Mutex<Vec<Session>>,
    pub current_session_id: Mutex<Option<String>>,
    /// 会话消息（降级路径：HTTP 不可用时前端仅读写内存态，此处提供消息查询能力）
    pub messages: Mutex<HashMap<String, Vec<Message>>>,
    /// 持久化文件路径（setup 时经 init_storage 绑定，None = 未持久化）
    storage_path: Mutex<Option<PathBuf>>,
}

impl Default for AppState {
    fn default() -> Self {
        Self {
            sessions: Mutex::new(Vec::new()),
            current_session_id: Mutex::new(None),
            messages: Mutex::new(HashMap::new()),
            storage_path: Mutex::new(None),
        }
    }
}

impl AppState {
    /// 绑定持久化文件路径并加载已有会话
    /// W8 修复：IPC 会话此前存内存 Vec，重启即清空——前端降级路径创建的会话
    /// 刷新/重启后消失（"数据丢失"）。现持久化到 app_data_dir/sessions.json。
    pub fn init_storage(&self, path: PathBuf) {
        if let Ok(content) = fs::read_to_string(&path) {
            match serde_json::from_str::<Vec<Session>>(&content) {
                Ok(loaded) => {
                    if let Ok(mut sessions) = self.sessions.lock() {
                        *sessions = loaded;
                    }
                }
                Err(e) => info!("sessions.json 解析失败，忽略并重建: {}", e),
            }
        }
        if let Ok(mut storage) = self.storage_path.lock() {
            *storage = Some(path);
        }
    }

    /// 将当前会话列表写回持久化文件（W8 修复）
    fn persist(&self) {
        let path = self.storage_path.lock().ok().and_then(|g| g.clone());
        let Some(path) = path else { return };
        let Ok(sessions) = self.sessions.lock() else { return };
        if let Ok(json) = serde_json::to_string_pretty(&*sessions) {
            if let Some(dir) = path.parent() {
                let _ = fs::create_dir_all(dir);
            }
            let _ = fs::write(&path, json);
        }
    }
}

#[tauri::command]
pub async fn list_sessions(
    state: State<'_, AppState>,
) -> Result<Vec<Session>, String> {
    info!("list_sessions called");
    let sessions = state.sessions.lock().map_err(|e| e.to_string())?;
    Ok(sessions.clone())
}

#[tauri::command]
pub async fn create_session(
    title: String,
    state: State<'_, AppState>,
) -> Result<Session, String> {
    info!("create_session called with title: {}", title);

    let now = chrono::Utc::now().timestamp_millis();
    let session = Session {
        id: Uuid::new_v4().to_string(),
        title,
        created_at: now,
        last_modified_at: now,
        message_count: 0,
    };

    let mut sessions = state.sessions.lock().map_err(|e| e.to_string())?;
    sessions.push(session.clone());

    let mut current_id = state.current_session_id.lock().map_err(|e| e.to_string())?;
    *current_id = Some(session.id.clone());

    // W8 修复：变更后落盘，重启保留
    state.persist();

    Ok(session)
}

#[tauri::command]
pub async fn switch_session(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    info!("switch_session called with id: {}", id);

    let sessions = state.sessions.lock().map_err(|e| e.to_string())?;
    let exists = sessions.iter().any(|s| s.id == id);

    if !exists {
        return Err("Session not found".to_string());
    }

    let mut current_id = state.current_session_id.lock().map_err(|e| e.to_string())?;
    *current_id = Some(id);

    Ok(())
}

#[tauri::command]
pub async fn delete_session(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    info!("delete_session called with id: {}", id);

    let mut sessions = state.sessions.lock().map_err(|e| e.to_string())?;
    sessions.retain(|s| s.id != id);

    let mut current_id = state.current_session_id.lock().map_err(|e| e.to_string())?;
    if *current_id == Some(id.clone()) {
        *current_id = sessions.first().map(|s| s.id.clone());
    }

    // W8 修复：变更后落盘，重启保留
    state.persist();

    Ok(())
}

#[tauri::command]
pub async fn get_current_session(
    state: State<'_, AppState>,
) -> Result<Option<Session>, String> {
    info!("get_current_session called");

    let current_id = state.current_session_id.lock().map_err(|e| e.to_string())?;
    let sessions = state.sessions.lock().map_err(|e| e.to_string())?;

    let session = current_id.as_ref().and_then(|id| {
        sessions.iter().find(|s| s.id == *id).cloned()
    });

    Ok(session)
}

#[tauri::command]
pub async fn rename_session(
    id: String,
    title: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    info!("rename_session called with id: {}, title: {}", id, title);

    let mut sessions = state.sessions.lock().map_err(|e| e.to_string())?;
    let session = sessions.iter_mut().find(|s| s.id == id);

    match session {
        Some(s) => {
            s.title = title;
            // W8 修复：变更后落盘，重启保留
            state.persist();
            Ok(())
        }
        None => Err("Session not found".to_string()),
    }
}

#[tauri::command]
pub async fn get_session(
    id: String,
    state: State<'_, AppState>,
) -> Result<Option<Session>, String> {
    info!("get_session called with id: {}", id);

    let sessions = state.sessions.lock().map_err(|e| e.to_string())?;
    Ok(sessions.iter().find(|s| s.id == id).cloned())
}

#[tauri::command]
pub async fn get_session_messages(
    session_id: String,
    state: State<'_, AppState>,
) -> Result<Vec<Message>, String> {
    info!("get_session_messages called with session_id: {}", session_id);

    let messages = state.messages.lock().map_err(|e| e.to_string())?;
    Ok(messages.get(&session_id).cloned().unwrap_or_default())
}

#[tauri::command]
pub async fn clear_all_sessions(
    state: State<'_, AppState>,
) -> Result<(), String> {
    info!("clear_all_sessions called");

    state.sessions.lock().map_err(|e| e.to_string())?.clear();
    *state
        .current_session_id
        .lock()
        .map_err(|e| e.to_string())? = None;
    state.messages.lock().map_err(|e| e.to_string())?.clear();
    // W8 修复：变更后落盘，重启保留
    state.persist();

    Ok(())
}

File: app/src/workspaces/commands/session.ts
// MIT License
// Copyright (c) 2026 190615273@qq.com
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

/**
 * 工作空间会话命令
 * 负责 Agent 会话级别的 Git Worktree 进入/退出
 * 同时暴露为 CLI 子命令 (workspace enter/exit) 和 Agent 工具 (EnterWorktreeTool/ExitWorktreeTool)
 */
import { execSync } from 'child_process';
import { existsSync } from 'fs';
import { join } from 'path';
import type { CommandContext, CommandResult } from '@modules/commands';
import { getLogger } from '@modules/monitoring';
const logger = getLogger('Workspace');

/**
 * 验证 worktree slug 格式
 * 只允许字母、数字、连字符和下划线
 */
function validateSlug(slug: string): boolean {
  return /^[a-zA-Z0-9_-]+$/.test(slug);
}

/**
 * 获取 Git 根目录
 */
function getGitRoot(cwd: string): string | null {
  try {
    const result = execSync('git rev-parse --show-toplevel', {
      encoding: 'utf8',
      stdio: 'pipe',
      cwd,
    });
    return result.trim();
  } catch {
    return null;
  }
}

/**
 * 进入 Worktree
 * 创建并切换到 git worktree，为 Agent 创建隔离的工作区
 * @param slug 工作区标识符
 * @param branch 基于的分支名称（可选）
 * @param cwd 当前工作目录
 */
export async function enterWorktree(
  slug: string,
  branch: string | undefined,
  cwd: string
): Promise<CommandResult> {
  if (!slug || slug.trim().length === 0) {
    return {
      success: false,
      type: 'error',
      error: 'slug 是必需的',
      message: '用法: /workspace enter <slug> [branch]',
    };
  }

  if (!validateSlug(slug)) {
    return {
      success: false,
      type: 'error',
      error: 'slug 格式无效。只允许字母、数字、连字符和下划线。',
      message: 'slug 格式无效。只允许字母、数字、连字符和下划线。',
    };
  }

  const gitRoot = getGitRoot(cwd);
  if (!gitRoot) {
    return {
      success: false,
      type: 'error',
      error: '当前目录不在 Git 仓库中',
      message: '当前目录不在 Git 仓库中。请在 Git 仓库内使用此命令。',
    };
  }

  const worktreePath = join(gitRoot, '..', `${slug}-worktree`);
  if (existsSync(worktreePath)) {
    return {
      success: false,
      type: 'error',
      error: `Worktree "${slug}" 已存在`,
      message: `Worktree "${slug}" 已存在于 ${worktreePath}`,
    };
  }

  const branchName = branch || `worktree/${slug}`;

  try {
    try {
      execSync(`git show-ref --verify --quiet refs/heads/${branchName}`, {
        stdio: 'pipe',
      });
    } catch {
      execSync(`git branch ${branchName}`, { stdio: 'pipe' });
    }

    execSync(`git worktree add "${worktreePath}" ${branchName}`, {
      stdio: 'pipe',
    });

    logger.info(`Worktree 已创建: ${slug} (${worktreePath})`);

    return {
      success: true,
      type: 'text',
      message: `Worktree "${slug}" 已创建在 ${worktreePath}，基于分支 ${branchName}`,
      data: { slug, worktree_path: worktreePath, branch: branchName },
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    return {
      success: false,
      type: 'error',
      error: `创建 worktree 失败: ${message}`,
      message: `创建 worktree 失败: ${message}`,
    };
  }
}

/**
 * 退出 Worktree
 * 退出 git worktree 并返回原分支，可选删除 worktree 目录
 * @param slug 工作区标识符
 * @param remove 是否删除 worktree 目录
 * @param cwd 当前工作目录
 */
export async function exitWorktree(
  slug: string,
  remove: boolean,
  cwd: string
): Promise<CommandResult> {
  if (!slug || slug.trim().length === 0) {
    return {
      success: false,
      type: 'error',
      error: 'slug 是必需的',
      message: '用法: /workspace exit <slug> [--remove]',
    };
  }

  try {
    const currentBranch = execSync('git branch --show-current', {
      encoding: 'utf8',
      stdio: 'pipe',
    }).trim();

    const isInWorktree =
      execSync('git rev-parse --is-inside-work-tree', {
        encoding: 'utf8',
        stdio: 'pipe',
      }).trim() === 'true';

    if (!isInWorktree) {
      return {
        success: false,
        type: 'error',
        error: '不在 git worktree 中',
        message: '当前不在 git worktree 中。',
      };
    }

    const worktreeList = execSync('git worktree list --porcelain', {
      encoding: 'utf8',
      stdio: 'pipe',
    });

    const worktreeLines = worktreeList.split('\n');
    let worktreePath = '';
    for (const line of worktreeLines) {
      if (line.startsWith('worktree ')) {
        worktreePath = line.replace('worktree ', '').trim();
        if (worktreePath.includes(slug)) {
          break;
        }
      }
    }

    if (!worktreePath || !existsSync(worktreePath)) {
      return {
        success: false,
        type: 'error',
        error: `Worktree "${slug}" 未找到`,
        message: `找不到 worktree "${slug}"。`,
      };
    }

    const gitRoot = getGitRoot(cwd);
    if (worktreePath === cwd && gitRoot) {
      process.chdir(gitRoot);
    }

    const flag = remove ? '' : '--force';
    execSync(`git worktree remove ${flag} "${worktreePath}"`.trim(), {
      stdio: 'pipe',
    });

    logger.info(`Worktree 已移除: ${slug}`);

    return {
      success: true,
      type: 'text',
      message: `Worktree "${slug}" 已移除。已返回到主分支。`,
      data: { slug, previous_branch: currentBranch },
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    return {
      success: false,
      type: 'error',
      error: `退出 worktree 失败: ${message}`,
      message: `退出 worktree 失败: ${message}`,
    };
  }
}

export default {
  /**
   * 执行会话子命令
   * @param subcommand 子命令名: enter | exit
   * @param args 命令参数
   * @param context 命令上下文
   */
  async execute(
    subcommand: string,
    args: string,
    context: CommandContext
  ): Promise<CommandResult> {
    const parts = args.trim().split(/\s+/);
    const slug = parts[0] || '';

    switch (subcommand) {
      case 'enter': {
        const branch = parts.slice(1).join(' ') || undefined;
        return await enterWorktree(slug, branch, context.cwd || process.cwd());
      }
      case 'exit': {
        const remove = args.includes('--remove');
        return await exitWorktree(slug, remove, context.cwd || process.cwd());
      }
      default:
        return {
          success: false,
          type: 'error',
          error: `未知子命令: ${subcommand}`,
          message: `未知子命令: workspace ${subcommand}`,
        };
    }
  },
};
